### PR TITLE
[MultiThreading] Parallel springs

### PR DESCRIPTION
**Pull request by alxbilger**
Based on https://github.com/sofa-framework/sofa/pull/3594

Introduce `ParallelStiffSpringForceField` and `ParallelMeshSpringForceField`

The base classes `SpringForceField` and `StiffSpringForceField` were also refactored to be able to reuse common code with their parallel implementations.

Only the `addForce` method is implemented. The provided example has an explicit integration scheme where the force derivative is not required. `addDForce` and `addKToMatrix` may come later.

Benchmark:

```
Sequential
1000 iterations done in 9.11327 s ( 109.73 FPS).

Parallel
1000 iterations done in 5.04016 s ( 198.406 FPS)
```

The implementation is naïve. We could probably do better. Note there are 11264 points and 62668 edges (springs) in the provided example.

______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

### CODE REVIEW

**Comment by alxbilger** [comment]
[ci-build][with-all-tests]

**Comment by alxbilger** [comment]
[ci-build][with-all-tests]

File: applications/plugins/MultiThreading/src/MultiThreading/component/solidmechanics/spring/ParallelStiffSpringForceField.inl [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    Inherit1::init();
+    initTaskScheduler();
+}
+
+template <class DataTypes>
+void ParallelStiffSpringForceField<DataTypes>::addForce(const sofa::core::MechanicalParams* mparams,
+    DataVecDeriv& data_f1, DataVecDeriv& data_f2, const DataVecCoord& data_x1,
+    const DataVecCoord& data_x2, const DataVecDeriv& data_v1, const DataVecDeriv& data_v2)
+{
+    VecDeriv&       f1 = *data_f1.beginEdit();
```
Comment (damienmarchal) [inline applications/plugins/MultiThreading/src/MultiThreading/component/solidmechanics/spring/ParallelStiffSpringForceField.inl]: Replace begin/end with getAccessor()

File: applications/plugins/MultiThreading/src/MultiThreading/component/solidmechanics/spring/ParallelMeshSpringForceField.inl [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+#include <MultiThreading/component/solidmechanics/spring/ParallelMeshSpringForceField.h>
+
+namespace multithreading::component::solidmechanics::spring
+{
+
+template <class DataTypes>
+void ParallelMeshSpringForceField<DataTypes>::init()
+{
+    MeshSpringForceField<DataTypes>::init();
```
Comment (damienmarchal) [inline applications/plugins/MultiThreading/src/MultiThreading/component/solidmechanics/spring/ParallelMeshSpringForceField.inl:32]: This is using the "call super" design pattern. Maybe a good idea not use this pattern anymore.

File: applications/plugins/MultiThreading/src/MultiThreading/component/solidmechanics/spring/ParallelStiffSpringForceField.inl (hunk 1 above)
Comment (alxbilger) [reply applications/plugins/MultiThreading/src/MultiThreading/component/solidmechanics/spring/ParallelStiffSpringForceField.inl]: done

**Comment by alxbilger** [comment]
[ci-build][with-all-tests]

**Comment by fredroy** [comment]
[ci-build][with-all-tests][force-full-build]

**Review by epernod — APPROVED** [review]

**Pull request merged**